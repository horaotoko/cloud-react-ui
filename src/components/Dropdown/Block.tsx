import styled from "styled-components";

export const Block = styled.div`
  padding: 16px;

  & + & {
    border-top: 1px solid #d4d8dd;
  }
`;

Block.displayName = "Dropdown.Block";
