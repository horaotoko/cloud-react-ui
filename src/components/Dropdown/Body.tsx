import styled from 'styled-components';

type DropdownBodyProps = {
  children?: React.ReactNode;
  width?: string;
  ariaHidden: boolean;
  placement?: 'left' | 'right';
};

export const Body = styled.div.attrs<DropdownBodyProps>(({ ariaHidden }) => ({
  'aria-hidden': ariaHidden,
}))<DropdownBodyProps>`
  width: ${({ width }) => width};
  position: absolute;
  top: calc(100% + 10px);
  left: ${({ placement }) => placement === 'left' && 0};
  right: ${({ placement }) => placement === 'right' && 0};
  transform-origin: top ${({ placement }) => placement};
  transition: opacity 299ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    transform 199ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  overflow-y: auto;
  opacity: 1;
  border-radius: 4px;
  background-color: #fff;
  box-shadow: 1px 2px 4px rgba(0, 0, 0, 0.3);
  visibility: visible;

  & > .block {
    padding: 8px 16px;

    & + & {
      border-top: 1px solid #d4d8dd;
    }
  }

  &[aria-hidden='true'] {
    transform: scale(0.75, 0.5625);
    transition: opacity 299ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
      transform 199ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
      visibility 299ms linear 299ms;
    visibility: hidden;
    opacity: 0;
  }
`;

Body.defaultProps = {
  width: 'max-content',
  placement: 'left',
};

Body.displayName = 'Dropdown.Body';
