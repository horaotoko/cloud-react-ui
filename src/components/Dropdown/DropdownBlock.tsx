import { forwardRef } from "react";
import styled, { css } from "styled-components";
import { defaultProps } from "../../theme";

export type DropdownBlockProps = {
  children?: React.ReactNode;
  className?: string;
  ariaLabel?: string;
  collapsed?: boolean;
  role?: string;
};

const StyledDropdownBlock = styled.div<DropdownBlockProps>`
  ${({ collapsed = false, theme: { dropdown } }) => css`
    padding: ${collapsed
      ? dropdown.block.padding.collapsed
      : dropdown.block.padding.default};
    & + & {
      border-top: ${dropdown.block.borderWidth} ${dropdown.block.borderStyle}
        ${dropdown.block.borderColor};
    }
  `}
`;
StyledDropdownBlock.defaultProps = defaultProps;

const DropdownBlock = forwardRef<HTMLDivElement, DropdownBlockProps>(
  (props, ref) => <StyledDropdownBlock ref={ref} {...props} />
);

DropdownBlock.displayName = "Dropdown.Block";

export { DropdownBlock };
