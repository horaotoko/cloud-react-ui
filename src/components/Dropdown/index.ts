export { Dropdown } from "./Dropdown";
export type {
  DropdownProps,
  DropdownBlockProps,
  DropdownBodyProps,
  DropdownItemProps,
} from "./Dropdown";
