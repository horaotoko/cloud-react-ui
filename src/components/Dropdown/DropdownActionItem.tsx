import { forwardRef } from "react";
import styled from "styled-components";
import { color, values, defaultProps } from "../../theme";
import { useDropdown } from "./Dropdown";

export type DropdownActionItemProps = {
  children?: React.ReactNode;
  className?: string;
  role?: string;
  ariaLabel?: string;
  isKeepOpen?: boolean;
};

const StyledDropdownActionItem = styled.div<DropdownActionItemProps>`
  & > a,
  & > button {
    width: 100%;
    display: flex;
    flex-direction: column;
    text-decoration: none;
    background-color: transparent;
    border: 0;
    padding: 8px 16px;
    line-height: 1;
    cursor: pointer;
    white-space: nowrap;
    user-select: none;
    appearance: none;
    font-family: inherit;
    font-size: ${values.text.size.middle};
    color: ${color.nightRider};
    &:hover,
    &:active,
    &:focus {
      color: ${color.royalBlue};
      text-decoration: none;
    }
  }
`;
StyledDropdownActionItem.defaultProps = defaultProps;

export const DropdownActionItem = forwardRef<
  HTMLDivElement,
  DropdownActionItemProps
>(({ isKeepOpen, ...props }, ref) => {
  const { setIsActive } = useDropdown();
  const handleClick = () => {
    if (isKeepOpen) {
      return;
    }
    setIsActive(false);
  };
  return (
    <StyledDropdownActionItem onClick={handleClick} ref={ref} {...props} />
  );
});

DropdownActionItem.displayName = "Dropdown.ActionItem";
