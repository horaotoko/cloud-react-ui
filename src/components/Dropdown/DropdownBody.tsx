import { forwardRef } from "react";
import styled, { css } from "styled-components";
import { values, color, zIndex, defaultProps } from "../../theme";
import { useDropdown } from "./Dropdown";

export type DropdownBodyProps = {
  children?: React.ReactNode;
  className?: string;
  ariaLabel?: string;
  width?: string;
  isOpen?: boolean;
  placement?: "left" | "right";
};

const StyledDropdownBody = styled.div.attrs<DropdownBodyProps>(
  ({ isOpen }) => ({
    "aria-hidden": !isOpen,
  })
)<DropdownBodyProps>`
  ${({ width = "max-content", placement = "left" }) => css`
    z-index: ${zIndex.dropdown};
    width: ${width};
    position: absolute;
    top: calc(100% + 10px);
    left: ${placement === "left" && 0};
    right: ${placement === "right" && 0};
    transform-origin: top ${placement};
    transition: opacity 299ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
      transform 199ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    overflow-y: auto;
    opacity: 1;
    border-radius: 4px;
    background-color: ${color.white};
    box-shadow: 1px 2px 4px rgba(0, 0, 0, 0.3);
    visibility: visible;

    & > .block {
      & + & {
        border-top: ${values.border.width.thin},
          ${values.border.style.solid} ${color.linkWater};
      }
    }

    &[aria-hidden="true"] {
      transform: scale(0.75, 0.5625);
      transition: opacity 299ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
        transform 199ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
        visibility 299ms linear 299ms;
      visibility: hidden;
      opacity: 0;
    }
  `}
`;
StyledDropdownBody.defaultProps = defaultProps;

export const DropdownBody = forwardRef<HTMLDivElement, DropdownBodyProps>(
  ({ isOpen, ...props }, ref) => {
    const { isActive } = useDropdown();
    return <StyledDropdownBody isOpen={isActive} ref={ref} {...props} />;
  }
);

DropdownBody.displayName = "Dropdown.Body";
