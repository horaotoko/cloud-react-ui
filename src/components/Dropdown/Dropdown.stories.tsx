import { Story, Meta } from "@storybook/react";
import { Dropdown, DropdownProps } from "./Dropdown";
import { Link } from "../../components";
import styled from "styled-components";

export default {
  component: Dropdown,
  title: "Components/Dropdown",
} as Meta;

const StyledLink = styled(Link)`
  display: flex;
  padding: 8px 16px;
`;

const Template: Story<DropdownProps> = (args) => (
  <Dropdown {...args}>
    <Dropdown.Block>
      <Dropdown.Item collapsed>
        <StyledLink href="#">ユーザー設定</StyledLink>
      </Dropdown.Item>
      <Dropdown.Item collapsed>
        <StyledLink href="#">ユーザー設定</StyledLink>
      </Dropdown.Item>
      <Dropdown.Item collapsed>
        <StyledLink href="#">ユーザー設定</StyledLink>
      </Dropdown.Item>
    </Dropdown.Block>
  </Dropdown>
);

export const Default = Template.bind({});
Default.args = { toggleLabel: "ラベル", width: "200px" };

export const Placement = Template.bind({});
Placement.args = {
  toggleLabel: "ラベル(placement: right)",
  placement: "right",
};

const MultipleTemplate: Story<DropdownProps> = (args) => (
  <Dropdown {...args}>
    <Dropdown.Block>
      <Dropdown.Item>
        Blockを複数配置することで、内容を分割できます。
      </Dropdown.Item>
    </Dropdown.Block>
    <Dropdown.Block>
      <Dropdown.Item collapsed>
        <StyledLink href="#">ユーザー設定</StyledLink>
      </Dropdown.Item>
      <Dropdown.Item collapsed>
        <StyledLink href="#">ユーザー設定</StyledLink>
      </Dropdown.Item>
      <Dropdown.Item collapsed>
        <StyledLink href="#">ユーザー設定</StyledLink>
      </Dropdown.Item>
    </Dropdown.Block>
    <Dropdown.Block collapsed>
      <Dropdown.Item collapsed>
        collapsedオプションがあれば余白を調整できます。
      </Dropdown.Item>
    </Dropdown.Block>
  </Dropdown>
);

export const MultipleBlock = MultipleTemplate.bind({});
MultipleBlock.args = { toggleLabel: "ラベル" };
