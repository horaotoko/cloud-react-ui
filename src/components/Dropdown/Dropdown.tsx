import { forwardRef } from "react";
import styled from "styled-components";
import { useState } from "react";
import { DropdownItem } from "./DropdownItem";
import { DropdownBlock } from "./DropdownBlock";
import { DropdownBody } from "./DropdownBody";
import { DropdownToggleButton } from "./DropdownToggleButton";

export type DropdownProps = {
  children?: React.ReactNode;
  width?: string;
  toggleLabel: string;
  placement?: "left" | "right";
};

const StyledDropdown = styled.div`
  position: relative;
  width: fit-content;
`;

const DropdownRoot = forwardRef<HTMLDivElement, DropdownProps>(
  ({ children, toggleLabel, width, placement }, ref) => {
    const [isOpen, setIsOpen] = useState(false);

    function handleClick() {
      setIsOpen(!isOpen);
    }

    return (
      <StyledDropdown ref={ref}>
        <DropdownToggleButton onClick={handleClick} ariaExpanded={isOpen}>
          {toggleLabel}
        </DropdownToggleButton>
        <DropdownBody width={width} ariaHidden={!isOpen} placement={placement}>
          {children}
        </DropdownBody>
      </StyledDropdown>
    );
  }
);

DropdownRoot.displayName = "Dropdown";

export type { DropdownBodyProps } from "./DropdownBody";
export type { DropdownBlockProps } from "./DropdownBlock";
export type { DropdownItemProps } from "./DropdownItem";
export const Dropdown = Object.assign(DropdownRoot, {
  Body: DropdownBody,
  Block: DropdownBlock,
  Item: DropdownItem,
});
