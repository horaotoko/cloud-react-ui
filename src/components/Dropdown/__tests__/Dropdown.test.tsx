import { render } from "@testing-library/react";
import { Dropdown, DropdownProps } from "../../Dropdown";
import { Link } from "../../../components";

describe("Dropdown", () => {
  const Template = (args: DropdownProps): JSX.Element => (
    <Dropdown {...args}>
      <Dropdown.Block>
        <Dropdown.Item>
          <Link href="#">ユーザー設定</Link>
        </Dropdown.Item>
        <Dropdown.Item>
          <Link href="#">ユーザー設定</Link>
        </Dropdown.Item>
        <Dropdown.Item>
          <Link href="#">ユーザー設定</Link>
        </Dropdown.Item>
      </Dropdown.Block>
    </Dropdown>
  );

  it("default", () => {
    const { asFragment } = render(<Template toggleLabel="ラベル" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("with placement", () => {
    const { asFragment } = render(
      <Template toggleLabel="ラベル" placement="right" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("with multiple blocks", () => {
    const { asFragment } = render(
      <Dropdown toggleLabel="ラベル">
        <Dropdown.Block>
          Blockを複数配置することで、内容を分割できます。
        </Dropdown.Block>
        <Dropdown.Block>
          <Dropdown.Item>
            <Link href="#">ユーザー設定</Link>
          </Dropdown.Item>
          <Dropdown.Item>
            <Link href="#">ユーザー設定</Link>
          </Dropdown.Item>
          <Dropdown.Item>
            <Link href="#">ユーザー設定</Link>
          </Dropdown.Item>
        </Dropdown.Block>
        <Dropdown.Block>
          Blockを複数配置することで、内容を分割できます。
        </Dropdown.Block>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
