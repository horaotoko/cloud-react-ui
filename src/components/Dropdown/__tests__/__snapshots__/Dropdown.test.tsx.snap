// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown default 1`] = `
<DocumentFragment>
  .c7 {
  font-size: 13px;
  padding: 8px 16px;
}

.c6 {
  padding: 8px 0;
}

.c5 + .c5 {
  border-top: 1px solid #D4D8DD;
}

.c4 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  position: absolute;
  top: calc(100% + 10px);
  left: 0;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, -webkit-transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  overflow-y: auto;
  opacity: 1;
  border-radius: 4px;
  background-color: #fff;
  box-shadow: 1px 2px 4px rgba(0,0,0,0.3);
  visibility: visible;
}

.c4 > .block {
  padding: 8px 16px;
}

.c3 > .block + .c3 > .block {
  border-top: 1px solid #D4D8DD;
}

.c4[aria-hidden="true"] {
  -webkit-transform: scale(0.75,0.5625);
  -ms-transform: scale(0.75,0.5625);
  transform: scale(0.75,0.5625);
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, -webkit-transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  visibility: hidden;
  opacity: 0;
}

.c2 {
  font-size: 13px;
  color: #333333;
}

.c2:hover {
  color: #333333;
}

.c8 {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:visited {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:focus {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:active {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:hover {
  cursor: pointer;
  color: #3B7DE9;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c8.disabled {
  cursor: not-allowed;
  color: #999999;
}

.c1 {
  font-size: 13px;
  color: #333333;
  background-color: #fff;
  border: none;
}

.c1 svg {
  margin-left: 6px;
}

.c0 {
  position: relative;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

<div
    class="c0"
  >
    <button
      aria-haspopup="true"
      class="c1"
    >
      ラベル
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-up fa-w-10 fa-rotate-180 c2"
        data-icon="caret-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        theme="[object Object]"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="c3 c4"
    >
      <div
        class="c5 c6"
      >
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown with multiple blocks 1`] = `
<DocumentFragment>
  .c7 {
  font-size: 13px;
  padding: 8px 16px;
}

.c6 {
  padding: 8px 0;
}

.c5 + .c5 {
  border-top: 1px solid #D4D8DD;
}

.c4 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  position: absolute;
  top: calc(100% + 10px);
  left: 0;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, -webkit-transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  overflow-y: auto;
  opacity: 1;
  border-radius: 4px;
  background-color: #fff;
  box-shadow: 1px 2px 4px rgba(0,0,0,0.3);
  visibility: visible;
}

.c4 > .block {
  padding: 8px 16px;
}

.c3 > .block + .c3 > .block {
  border-top: 1px solid #D4D8DD;
}

.c4[aria-hidden="true"] {
  -webkit-transform: scale(0.75,0.5625);
  -ms-transform: scale(0.75,0.5625);
  transform: scale(0.75,0.5625);
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, -webkit-transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  visibility: hidden;
  opacity: 0;
}

.c2 {
  font-size: 13px;
  color: #333333;
}

.c2:hover {
  color: #333333;
}

.c8 {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:visited {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:focus {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:active {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:hover {
  cursor: pointer;
  color: #3B7DE9;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c8.disabled {
  cursor: not-allowed;
  color: #999999;
}

.c1 {
  font-size: 13px;
  color: #333333;
  background-color: #fff;
  border: none;
}

.c1 svg {
  margin-left: 6px;
}

.c0 {
  position: relative;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

<div
    class="c0"
  >
    <button
      aria-haspopup="true"
      class="c1"
    >
      ラベル
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-up fa-w-10 fa-rotate-180 c2"
        data-icon="caret-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        theme="[object Object]"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="c3 c4"
    >
      <div
        class="c5 c6"
      >
        Blockを複数配置することで、内容を分割できます。
      </div>
      <div
        class="c5 c6"
      >
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
      </div>
      <div
        class="c5 c6"
      >
        Blockを複数配置することで、内容を分割できます。
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown with placement 1`] = `
<DocumentFragment>
  .c7 {
  font-size: 13px;
  padding: 8px 16px;
}

.c6 {
  padding: 8px 0;
}

.c5 + .c5 {
  border-top: 1px solid #D4D8DD;
}

.c4 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  position: absolute;
  top: calc(100% + 10px);
  right: 0;
  -webkit-transform-origin: top right;
  -ms-transform-origin: top right;
  transform-origin: top right;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, -webkit-transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms;
  overflow-y: auto;
  opacity: 1;
  border-radius: 4px;
  background-color: #fff;
  box-shadow: 1px 2px 4px rgba(0,0,0,0.3);
  visibility: visible;
}

.c4 > .block {
  padding: 8px 16px;
}

.c3 > .block + .c3 > .block {
  border-top: 1px solid #D4D8DD;
}

.c4[aria-hidden="true"] {
  -webkit-transform: scale(0.75,0.5625);
  -ms-transform: scale(0.75,0.5625);
  transform: scale(0.75,0.5625);
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, -webkit-transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  -webkit-transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  transition: opacity 299ms cubic-bezier(0.4,0,0.2,1) 0ms, transform 199ms cubic-bezier(0.4,0,0.2,1) 0ms, visibility 299ms linear 299ms;
  visibility: hidden;
  opacity: 0;
}

.c2 {
  font-size: 13px;
  color: #333333;
}

.c2:hover {
  color: #333333;
}

.c8 {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:visited {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:focus {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:active {
  color: #3B7DE9;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:hover {
  cursor: pointer;
  color: #3B7DE9;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c8.disabled {
  cursor: not-allowed;
  color: #999999;
}

.c1 {
  font-size: 13px;
  color: #333333;
  background-color: #fff;
  border: none;
}

.c1 svg {
  margin-left: 6px;
}

.c0 {
  position: relative;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

<div
    class="c0"
  >
    <button
      aria-haspopup="true"
      class="c1"
    >
      ラベル
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-up fa-w-10 fa-rotate-180 c2"
        data-icon="caret-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        theme="[object Object]"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="c3 c4"
    >
      <div
        class="c5 c6"
      >
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
        <div
          class="c7"
        >
          <a
            class="c8"
            href="#"
          >
            ユーザー設定
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
